softioc/autosave.py

```diff
@@ -0,0 +1,274 @@
+import atexit
+import sys
+import threading
+import traceback
+from datetime import datetime
+from os import rename
+from pathlib import Path
+from shutil import copy2
+
+import numpy
+import yaml
+
+from .device_core import LookupRecordList
+
+SAV_SUFFIX = "softsav"
+SAVB_SUFFIX = "softsavB"
+DEFAULT_SAVE_PERIOD = 30.0
+
+
+def _ndarray_representer(dumper, array):
+    return dumper.represent_sequence(
+        "tag:yaml.org,2002:seq", array.tolist(), flow_style=True
+    )
+
+yaml.add_representer(numpy.ndarray, _ndarray_representer, Dumper=yaml.Dumper)
+
+def configure(
+    directory,
+    name,
+    save_period=DEFAULT_SAVE_PERIOD,
+    timestamped_backups=True,
+    enabled=True
+):
+    """This should be called before initialising the IOC. Configures the
+    autosave thread for periodic backing up of PV values.
+
+    Args:
+        directory: string or Path giving directory path where autosave backup
+            files are saved and loaded.
+        name: string name of the root used for naming backup files, this
+            is usually the same as the device name.
+        save_period: time in seconds between backups. Backups are only performed
+            if PV values have changed.
+        timestamped_backups: boolean which determines if backups of existing
+            autosave files are timestamped on IOC restart. True by default, if
+            False then backups get overwritten on each IOC restart.
+        enabled: boolean which enables or disables autosave, set to True by
+            default, or False if configure not called.
+    """
+    Autosave.directory = Path(directory)
+    Autosave.timestamped_backups = timestamped_backups
+    Autosave.save_period = save_period
+    Autosave.enabled = enabled
+    Autosave.device_name = name
+
+
+def start_autosave_thread():
+    worker = threading.Thread(
+        target=Autosave._loop,
+    )
+    worker.daemon = True
```

Comment:
After a chat with @Araneidae we think it may not be necessary to use a daemon thread. You already have a clean shutdown mechanism when using the Event, so by removing this line we should avoid the potential for file corruption on shutdown.